### ISSUE TITLE
Pose3SLAM benchmark nondeterministic

### ISSUE THREAD

**Issue by marcrasi** [borglab/SwiftFusion#88]
I noticed that the Pose3SLAM benchmark gives me different final losses every time that I run it. As far as I know, all the computations should be deterministic.

We should get to the bottom of this because it might indicate a bug. For example, the generic factor graph datastructure uses unsafe memory operations, and we might be reading uninitialized memory somewhere.

**Comment by ProfFan** [comment]
I agree on that, and I observed that for some difficult datasets, the nonlinear loss sometimes go to `nan` which indicates possibly a division-by-zero or something. I suspect a impl error in our Pose3 but I am not super sure. Is there a way to raise an exception when we have `nan`?

**Comment by marcrasi** [comment]
> Is there a way to raise an exception when we have `nan`?

I don't think there's an automatic way to do this.

Manually adding checks to the initializers of all our math types might be a good approach though. It seems like this would catch problems close to where they occur. The check should probably use a function that we define that is controlled by an `#if` flag -- `assert` isn't suitable because it gets disabled in `-O` mode and I thing we'll want to be able to debug nans in `-O` mode.
